### PR TITLE
`fixNonFungibleTokensV1_2`: Multiple bug fixes for XLS-20 NFTs

### PR DESCRIPTION
**Pull request by kennyzlei**
## High Level Overview of Change
Proposal to merge feature branch `feature/nft-fixes` to `develop` branch

### Context of Change
Introduce a new amendment, **`fixNonFungibleTokensV1_2`** which is a combination of bug fixes that have been individually merged into `feature/nft-fixes` through the pull request process:
- https://github.com/XRPLF/rippled/pull/4346
- https://github.com/XRPLF/rippled/pull/4380
- https://github.com/XRPLF/rippled/pull/4403
- https://github.com/XRPLF/rippled/pull/4399

Each individual pull request contains details on the bugs they intend to fix

### Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### CODE REVIEW

**Comment by kennyzlei** [comment]
@ximinez @intelliot I noticed the `feature/nft-fixes` branch is currently about 8 commits behind `develop` and causing a merge conflict.  The resolution itself is a few lines of code to catch up on other features added from what I have tested.  I assume `feature/nft-fixes` branch is protected and there are limitations to pushing commits to the branch directly to fix it.  Would appreciate some recommendations on getting it to a good state.  I was thinking these were some options:
1. Using an existing PR or new PR to pull in the ~8 commits into `feature/nft-fixes`.  Might be a bit strange to have duplicate diff in that PR but it'll get the branch up-to-date
2. Have a maintainer or someone with elevated perms push a commit directly to the `feature/nft-fixes` branch to fix the conflict

**Comment by intelliot** [comment]
I don't think `feature/nft-fixes` is protected. If we bring it up-to-date with a merge commit, there will not be any duplication in the diff.

File: src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp
Comment (nbougalis) [inline src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp:280]: Note: negative balances are *not* a problem. Trustlines are bi-directional, and the issuer *might* have configured the trustline so that you can go into debt. The problem is a balance that _exceeds_ the limit.

Comment (intelliot) [reply src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp:300]: (@ledhed2222 wrote this comment [here](https://github.com/ledhed2222/rippled/commit/36c87451275ebd4dfdbb7972d5460555e7b49faf))

@ledhed2222 if it makes sense to you, could you clarify this comment? You can do so in a new PR with `feature/nft-fixes`.

(or, you can push a commit on `feature/nft-fixes`, if permissions allow. Either way is fine)

**Comment by scottschurr** [comment]
Just as a note, some of the the commit messages on this pull request do not follow guidelines we've been following for a long time.  Here's the source of the guidelines we've followed in the past: https://cbea.ms/git-commit/

Specifically, we've requested that line lengths on commit messages be limited to 72 characters or so.  We haven't always been great at enforcing that, but it has been a guideline.

I understand that expectations change over time.  So if unlimited line lengths is the new norm, then so be it.  But I thought ya'll should be aware that it is a change.

**Comment by intelliot** [comment]
It's a good guideline, but writing good (short) commit messages is extremely difficult, and I wouldn't hold up the release just for that. To take a specific example, how would you rewrite `[Amendment] If you set a destination on an NFT offer, only that destination can settle through brokerage (fix #4373) #4399`?

It is the worst offender and I would be in favor of shortening it with a simple rebase before merging it into `develop`. However, in order to do that, we need a specific subject line to use.

**Comment by ximinez** [comment]
> It's a good guideline, but writing good (short) commit messages is extremely difficult, and I wouldn't hold up the release just for that. To take a specific example, how would you rewrite `[Amendment] If you set a destination on an NFT offer, only that destination can settle through brokerage (fix #4373) (#4399)`?

```
Only account specified as destination can settle through brokerage:

* Fixes 4373
```
First line is 68 characters long. (And I'd drop `[Amendment]` from all the commit messages.)

**Comment by scottschurr** [comment]
@intelliot, there are two guidelines I'm noticing:

1. The subject line should be limited to 50 characters and
2. The lines in the body should be limited to 72 characters.

You're noting a place where we broke the first of these guidelines.  50 characters is really hard.  I tend to target 60 characters.  For the case you pointer out I might suggest that we don't need all of those detail in the subject line.  We can put additional details in the body:

```
Place new restrictions on brokering NFT offers:

When brokering NFT offers, if there's a destination field in the offer
we now require that the broker must be the destination.
```

These guidelines are hard.  Even native English speaking programmers tend to not be great writers.  Placing these constraints on non-native-English speakers is even harder.  But the guidelines do help a great deal with the legibility of the log.

**Comment by intelliot** [comment]
> > It's a good guideline, but writing good (short) commit messages is extremely difficult, and I wouldn't hold up the release just for that. To take a specific example, how would you rewrite `[Amendment] If you set a destination on an NFT offer, only that destination can settle through brokerage (fix #4373) (#4399)`?
> 
> ```
> Only account specified as destination can settle through brokerage:
> 
> * Fixes 4373
> ```
> 
> First line is 68 characters long. (And I'd drop `[Amendment]` from all the commit messages.)

To make it easier to navigate between the [commit history](https://github.com/XRPLF/rippled/commits/develop) and the relevant PRs, I'd like to include the PR # in the subject. Also, I believe putting the `#` in front of the issue number will make a link in the GitHub UI:

```
Only account specified as destination can settle through brokerage: (#4399)

* Fixes #4373
```

Unfortunately, that is now 75 characters

**Comment by kennyzlei** [comment]
I'll do that cleanup to remove `[Amendment]` and rename that particular commit to `Only account specified as destination can settle through brokerage: (#4399)`

**Comment by intelliot** [comment]
@kennyzlei Given you're rebasing anyway, and given that we would like to merge this PR as multiple commits (not a single squashed commit), feel free to rebase on top of the current HEAD of `develop`.

**Comment by kennyzlei** [comment]
@intelliot @scottschurr @ximinez just rebased and fixed up commit messages, ready for review

**Review by shawnxie999 — APPROVED** [review]

**Review by thejohnfreeman — APPROVED** [review]
I have not reviewed the code itself, but I confirmed that this is a feature branch whose individual commits were previously approved by trusted reviewers.

**Comment by intelliot** [comment]
(Scott S is reviewing)

**Review by scottschurr — APPROVED** [review]
Looks like what I'd expect.

**Review by intelliot — APPROVED** [review]

**Pull request merged**
